GAIZKIA ADELINE ATMAKA


2501972493 LA05


FINAL EXAM DEEP LEARNING NO. 2 (RoBERTa)

In [ ]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from sklearn.metrics import classification_report
from sklearn.metrics import precision_score, recall_score, f1_score, accuracy_score

Import Dataset

In [ ]:
df = pd.read_csv('Emotion.csv')
df

text    label
0                                i didnt feel humiliated  sadness
1      i can go from feeling so hopeless to so damned...  sadness
2       im grabbing a minute to post i feel greedy wrong    anger
3      i am ever feeling nostalgic about the fireplac...     love
4                                   i am feeling grouchy    anger
...                                                  ...      ...
19995  im having ssa examination tomorrow in the morn...  sadness
19996  i constantly worry about their fight against n...      joy
19997  i feel its important to share this info for th...      joy
19998  i truly feel that if you are passionate enough...      joy
19999  i feel like i just wanna buy any cute make up ...      joy

[20000 rows x 2 columns]

Check Informasi dan jumlah label dari dataset

In [ ]:
print(df.info())

print(df['label'].value_counts())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20000 entries, 0 to 19999
Data columns (total 2 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   text    20000 non-null  object
 1   label   20000 non-null  object
dtypes: object(2)
memory usage: 312.6+ KB
None
joy         6761
sadness     5797
anger       2709
fear        2373
love        1641
surprise     719
Name: label, dtype: int64


Tidak ada data kolom yang kosong didalam dataset tersebut. Ada 6 kelas pada kolom label

Lakukan preprocessing & Encoding untuk mengubah value label menjadi numeric

In [ ]:
label = {'sadness': 0, 'anger': 1, 'love': 2, 'surprise': 3, 'fear': 4, 'joy': 5}

X = df['text'].values
y = df['label'].values

y_encode = [label[i] for i in df['label']]
y_encode

[0,
 0,
 1,
 2,
 1,
 0,
 3,
 4,
 5,
 2,
 0,
 5,
 1,
 0,
 5,
 5,
 0,
 0,
 0,
 4,
 1,
 4,
 5,
 5,
 1,
 0,
 0,
 0,
 1,
 5,
 5,
 4,
 3,
 1,
 5,
 5,
 5,
 5,
 1,
 5,
 5,
 5,
 5,
 5,
 0,
 0,
 5,
 2,
 5,
 1,
 5,
 0,
 1,
 4,
 5,
 0,
 0,
 3,
 5,
 5,
 5,
 2,
 4,
 4,
 3,
 1,
 1,
 0,
 2,
 5,
 0,
 0,
 5,
 0,
 0,
 0,
 5,
 5,
 5,
 1,
 0,
 1,
 1,
 1,
 5,
 5,
 5,
 5,
 0,
 4,
 2,
 1,
 0,
 1,
 2,
 0,
 5,
 5,
 0,
 1,
 2,
 5,
 5,
 5,
 0,
 5,
 5,
 5,
 5,
 0,
 5,
 5,
 2,
 0,
 0,
 5,
 5,
 0,
 5,
 5,
 4,
 4,
 4,
 0,
 2,
 5,
 5,
 2,
 4,
 3,
 5,
 5,
 5,
 5,
 1,
 4,
 5,
 1,
 2,
 1,
 0,
 5,
 0,
 1,
 5,
 3,
 0,
 1,
 1,
 0,
 5,
 4,
 5,
 5,
 4,
 0,
 3,
 3,
 5,
 1,
 4,
 1,
 0,
 1,
 0,
 4,
 0,
 5,
 3,
 4,
 5,
 1,
 5,
 1,
 5,
 4,
 4,
 0,
 5,
 5,
 0,
 3,
 5,
 4,
 5,
 0,
 5,
 5,
 5,
 4,
 5,
 3,
 5,
 1,
 0,
 0,
 5,
 1,
 0,
 5,
 5,
 3,
 5,
 4,
 5,
 4,
 0,
 4,
 2,
 0,
 4,
 2,
 0,
 0,
 1,
 5,
 2,
 1,
 0,
 3,
 1,
 5,
 0,
 1,
 5,
 5,
 5,
 5,
 5,
 0,
 5,
 2,
 5,
 0,
 1,
 3,
 5,
 1,
 5,
 2,
 0,
 0,
 5,
 0,
 0,
 5,
 0,
 5,
 5,
 5,


Split the data into training 70%, validation 15% , and test sets 15%

In [ ]:
train_size = int(len(X) * 0.7)
val_size = int(len(X) * 0.15)
test_size = len(X) - (train_size - val_size)

X_train, y_train = X[:train_size], y_encode[:train_size]
X_val, y_val = X[train_size:train_size + val_size], y_encode[train_size + val_size:]
X_test, y_test = X[train_size + val_size:], y_encode[train_size + val_size:]

Instal Library yang dibutuhkan

In [ ]:
pip install --upgrade keras-nlp

Load RoBERTa preprocessor & Create model (RoBERTa)

In [ ]:
import keras_nlp

# print(keras_nlp.models.RobertaPreprocessor.presets.keys())
preprocessor = keras_nlp.models.RobertaPreprocessor.from_preset(
    "roberta_base_en",
    sequence_length=100
)

model = keras_nlp.models.RobertaClassifier.from_preset(
    "roberta_base_en",
    num_classes=6,
    preprocessor=preprocessor,
    activation="softmax"
)

Using TensorFlow backend


100%|██████████| 463/463 [00:00<00:00, 148kB/s]
100%|██████████| 446k/446k [00:00<00:00, 530kB/s]
100%|██████████| 0.99M/0.99M [00:01<00:00, 1.02MB/s]
100%|██████████| 498/498 [00:00<00:00, 504kB/s]
100%|██████████| 474M/474M [00:32<00:00, 15.4MB/s]
/usr/local/lib/python3.10/dist-packages/keras_nlp/src/models/backbone.py:37: UserWarning: `Model.state_updates` will be removed in a future version. This property should not be used in TensorFlow 2.0, as `updates` are applied automatically.
  return id(getattr(self, attr)) not in self._functional_layer_ids
/usr/local/lib/python3.10/dist-packages/keras_nlp/src/models/backbone.py:37: UserWarning: `layer.updates` will be removed in a future version. This property should not be used in TensorFlow 2.0, as `updates` are applied automatically.
  return id(getattr(self, attr)) not in self._functional_layer_ids


Train model

In [ ]:
model.fit(x = X_train, y = y_train, validation_data = (X_val, y_val), batch_size = 32, epochs = 5)

Epoch 1/5
438/438 [==============================] - 416s 752ms/step - loss: 0.5666 - sparse_categorical_accuracy: 0.7976 - val_loss: 4.4579 - val_sparse_categorical_accuracy: 0.2300
Epoch 2/5
438/438 [==============================] - 316s 722ms/step - loss: 0.1725 - sparse_categorical_accuracy: 0.9307 - val_loss: 5.1798 - val_sparse_categorical_accuracy: 0.2267
Epoch 3/5
438/438 [==============================] - 262s 598ms/step - loss: 0.1258 - sparse_categorical_accuracy: 0.9443 - val_loss: 5.6560 - val_sparse_categorical_accuracy: 0.2313
Epoch 4/5
438/438 [==============================] - 265s 604ms/step - loss: 0.1107 - sparse_categorical_accuracy: 0.9471 - val_loss: 5.7374 - val_sparse_categorical_accuracy: 0.2273
Epoch 5/5
438/438 [==============================] - 263s 601ms/step - loss: 0.0958 - sparse_categorical_accuracy: 0.9523 - val_loss: 6.2082 - val_sparse_categorical_accuracy: 0.2357


Untuk membuat model RoBERTa kita akan menggunakan function RobertaPreprocessor.

Parameter yang saya masukkan adalah:
Nama model yang dipakai, num_classes yang saya set 6 karena ada 6 kelas yang akan diklasifikasi, preprocessor yang berfungsi untuk menjalankan input saat nanti dilakukan fit.

Dari hasil training didapatkan accuracy yang didapatkan baik sebesar 0.9523 untuk training, namun pada validation didapatkan accuracy yang jelek sebesar 0.2357. Pada loss didapatkan sebesar 0.0958 untuk training dan pada validation didapatkan sebesar 6.2082

Dapat disimpulkan bahwa Model mengalami overfitting karena model berusaha menyempurnakan performanya pada training (mendapatkan accuracy yg sangat baik) namun pada val performanya sangat jelek karena terlalu fokus dalam menangkap pola yang ada pada train data.

Performance Analysis (precision, recall, F1-Score and accuracy)

In [ ]:
y_pred = model.predict(X_test)
y_pred = np.argmax(y_pred, axis = 1)

print(classification_report(y_test, y_pred))

94/94 [==============================] - 33s 275ms/step
              precision    recall  f1-score   support

           0       0.95      0.98      0.97       836
           1       0.93      0.94      0.94       410
           2       0.86      0.84      0.85       254
           3       1.00      0.68      0.81       117
           4       0.91      0.89      0.90       321
           5       0.94      0.96      0.95      1062

    accuracy                           0.94      3000
   macro avg       0.93      0.88      0.90      3000
weighted avg       0.94      0.94      0.93      3000



In [ ]:
print(precision_score(y_test, y_pred, average = 'macro'))
print(recall_score(y_test, y_pred, average = 'macro'))
print(f1_score(y_test, y_pred, average = 'macro'))
print(accuracy_score(y_test, y_pred))

0.9321169084555404
0.8815782030439832
0.9015285234821006
0.935


Dapat dilihat dari hasil analysis nilai precision, accuracy, recall, F1-Score sangat tinggi. Hal itu menandakan bahwa model RoBERTa dapat melakukan klasifikasi dengan sangat baik.


Model RoBERTa mampu memprediksi nilai positif dengan sangat akurat, berdasarkan nilai precision dan recall yang tinggi. Nilai F1-Score yang tinggi menunjukkan bahwa model mampu menyeimbangkan antara akurasi dan sensitivitas. Nilai accuracy yang tinggi menunjukkan bahwa model mampu melakukan tugas klasifikasi dengan baik test data.

***What is the difference between BERT and GPT? ***


BERT berfokus dalam memahami hubungan antar kata dalam kalimat, dimana dia akan menganalisis teks secara 2 arah yaitu dengan melihat kata sebelum dan sesudah. Sehingga BERT unggul dalam pemahaman makna dan cocok dipakai untuk menjawab pertanyaan dan klasifikasi teks.


GPT berfokus pada kekreatifan dalam menghasilkan teks dengan mengandalkan kata sebelumnya untuk memprediksi kata berikutnya. GPT cocok dipakai untuk menulis cerita atau terjemahan karena pandai merangkai kalimat sendiri.